### PR TITLE
feat: refactor mcore train/forward utilities

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

Nightly test results:

```
============================================================
sft-llama3.1-8b-1n8g-megatron
============================================================
                                 Metric Checks
┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┓
┃ Status ┃ Check                                ┃ Value              ┃ Message ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━┩
│ PASS   │ data["train/loss"]["1"] < 0.6        │ 0.5873891115188599 │         │
│ PASS   │ data["train/loss"]["250"] < 0.36     │ 0.3181086480617523 │         │
│ PASS   │ mean(data["timing/train/total_step_… │ 12.973950623508438 │         │
│        │ 2) < 20                              │                    │         │
└────────┴──────────────────────────────────────┴────────────────────┴─────────┘

============================================================
sft-llama3.1-8b-1n8g-megatron-lora
============================================================
                                 Metric Checks
┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┓
┃ Status ┃ Check                                ┃ Value              ┃ Message ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━┩
│ PASS   │ data["train/loss"]["1"] < 1.0        │ 0.8588194251060486 │         │
│ PASS   │ data["train/loss"]["50"] < 0.8       │ 0.717399537563324  │         │
│ PASS   │ max(data["ray/node.0.gpu.0.mem_gb"]) │ 52.1015625         │         │
│        │ < 60                                 │                    │         │
│ PASS   │ mean(data["timing/train/total_step_… │ 22.606327738080704 │         │
│        │ 2) < 30                              │                    │         │
└────────┴──────────────────────────────────────┴────────────────────┴─────────┘

============================================================
sft-llama3.1-8b-1n8g-megatron-seqpack
============================================================
                                 Metric Checks
┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┓
┃ Status ┃ Check                                ┃ Value              ┃ Message ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━┩
│ PASS   │ data["train/loss"]["1"] < 0.6        │ 0.5874463319778442 │         │
│ PASS   │ data["train/loss"]["250"] < 0.36     │ 0.3183392286300659 │         │
│ PASS   │ mean(data["timing/train/total_step_… │ 5.329389685129066  │         │
│        │ 2) < 6                               │                    │         │
└────────┴──────────────────────────────────────┴────────────────────┴─────────┘

============================================================
sft-qwen2.5-math7b-2n8g-megatron
============================================================
                                 Metric Checks
┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┓
┃ Status ┃ Check                               ┃ Value               ┃ Message ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━┩
│ PASS   │ data["train/loss"]["80"] < 0.301    │ 0.29687559604644775 │         │
│ PASS   │ data["validation/val_loss"]["80"] < │ 0.30329111218452454 │         │
│        │ 0.304                               │                     │         │
└────────┴─────────────────────────────────────┴─────────────────────┴─────────┘
```
```
============================================================
dpo-llama3.1-8b-instruct-4n8g-megatrontp2pp2-quick
============================================================
                                 Metric Checks
┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┓
┃ Status ┃ Check                                ┃ Value              ┃ Message ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━┩
│ PASS   │ data["train/loss"]["1"] < 3.6        │ 3.5130887031555176 │         │
│ PASS   │ data["train/loss"]["20"] < 3.4       │ 3.2065513134002686 │         │
│ PASS   │ data["train/preference_loss"]["1"] > │ 0.6931471824645996 │         │
│        │ 0.69314                              │                    │         │
│ PASS   │ data["train/preference_loss"]["1"] < │ 0.6931471824645996 │         │
│        │ 0.69316                              │                    │         │
│ PASS   │ data["train/preference_loss"]["20"]  │ 0.5910405516624451 │         │
│        │ < 0.6                                │                    │         │
│ PASS   │ mean(data["timing/train/total_step_… │ 4.939984655380249  │         │
│        │ -10) < 6.7                           │                    │         │
└────────┴──────────────────────────────────────┴────────────────────┴─────────┘
```

# Issues
Closes https://github.com/NVIDIA-NeMo/RL/issues/1593.
Closes https://github.com/NVIDIA-NeMo/RL/issues/1744.

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Added post-processing utilities for loss computation, log probability generation, and top-k logits extraction.
  * Introduced pipeline-parallel broadcasting utilities for distributed training.
  * Enhanced training infrastructure with improved forward/backward pass orchestration.

* **Refactor**
  * Streamlined training workflow with consolidated post-processing architecture.
  * Refactored data handling to support improved type safety and distributed operations.

* **Tests**
  * Added comprehensive unit test coverage for training utilities and post-processing pipelines.

### CODE REVIEW

File: nemo_rl/models/megatron/common.py [hunk 1]
```diff
@@ -42,119 +31,6 @@ def _round_up_to_multiple(value: int, multiple: int) -> int:
...
-
-    additional_kwargs = {}
-    # Mamba models currently do not support packed_seq_params
-    if packed_seq_params is not None:
-        additional_kwargs["packed_seq_params"] = packed_seq_params
-
-    if defer_fp32_logits:
-        additional_kwargs["fp32_output"] = False
-
-    with straggler_timer:
```
Comment (cuichenx) [inline nemo_rl/models/megatron/common.py:106]: not too familiar with this, but the straggler timer is removed and not added back elsewhere, is this expected?

File: nemo_rl/models/megatron/train.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        This function wraps a loss function with the necessary context and parameters
+        to compute loss and metrics from model outputs. It handles sequence packing
+        and context parallelism normalization.
+
+        Args:
+            loss_fn: The base loss function to wrap
+            cfg (dict): Configuration dictionary
+            data_dict: Batched data dictionary
+            packed_seq_params: Parameters for packed sequences (optional)
+            cp_normalize (bool): Whether to normalize by context parallel size
```
Comment (cuichenx) [inline nemo_rl/models/megatron/train.py]: the docstring doesn't match function signature here

File: nemo_rl/models/megatron/train.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+    if len(multimodal_data) > 0:
+        position_ids = None
+
+    additional_kwargs = {}
+    # Mamba models currently do not support packed_seq_params
+    if packed_seq_params is not None:
+        additional_kwargs["packed_seq_params"] = packed_seq_params
+    if defer_fp32_logits:
+        additional_kwargs["fp32_output"] = False
+    # with straggler_timer:
```
Comment (terrykong) [inline nemo_rl/models/megatron/train.py]: same Q about the straggler timer

File: nemo_rl/models/megatron/data.py [hunk 4]
```diff
@@ -211,105 +201,78 @@ def process_microbatch(
...
-        if pack_sequences:
-            # For packed sequences with padded input, we need sequence lengths
-            assert seq_length_key is not None, (
-                "seq_length_key must be provided for packed sequences"
-            )
-            assert seq_length_key in data_dict, (
-                f"{seq_length_key} not found in data_dict"
-            )
+):
+    # with straggler_timer(bdata=True):
```
Comment (terrykong) [inline nemo_rl/models/megatron/data.py]: same q about straggler timer

File: nemo_rl/models/megatron/train.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+        input_ids=input_ids_cp_sharded,
+        position_ids=position_ids,
+        attention_mask=attention_mask,
+        **additional_kwargs,
+        **multimodal_data,
+    )
+
+    # Apply temperature scaling to logits for training
+    # This matches the dtensor worker's _apply_temperature_scaling in the train method
+    if "generation" in cfg and cfg["generation"] is not None:
```
Comment (terrykong) [inline nemo_rl/models/megatron/train.py]: for consistency, shall we also have the same `apply_temperature_scaling()` https://github.com/NVIDIA-NeMo/RL/pull/1757/changes#diff-b193ca47f5466d66a35db24b0d7f1c96dca88663e65e938bba8837cec310d758R137

File: nemo_rl/models/megatron/train.py [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+
+    return output_tensor
+
+
+def forward_with_post_processing_fn(
+    data_iterator: Iterator[ProcessedMicrobatch],
+    model: GPTModel,
+    cfg: Dict[str, Any],
+    post_processing_fn: PostProcessingFunction,
+    defer_fp32_logits: Optional[bool] = True,
```
Comment (terrykong) [inline nemo_rl/models/megatron/train.py]: false for consistency?

File: nemo_rl/models/megatron/train.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+    if "generation" in cfg and cfg["generation"] is not None:
+        output_tensor.div_(cfg["generation"]["temperature"])
+
+    return output_tensor
+
+
+def forward_with_post_processing_fn(
+    data_iterator: Iterator[ProcessedMicrobatch],
+    model: GPTModel,
+    cfg: Dict[str, Any],
```
Comment (terrykong) [inline nemo_rl/models/megatron/train.py]: same question on the automodel PR, but can we narrow this to the policyconfig?

File: nemo_rl/models/megatron/train.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    # Apply temperature scaling to logits for training
+    # This matches the dtensor worker's _apply_temperature_scaling in the train method
+    if "generation" in cfg and cfg["generation"] is not None:
+        output_tensor.div_(cfg["generation"]["temperature"])
+
+    return output_tensor
+
+
+def forward_with_post_processing_fn(
+    data_iterator: Iterator[ProcessedMicrobatch],
```
Comment (terrykong) [inline nemo_rl/models/megatron/train.py:122]: is it possible to make this consistent with the automodel path that takes a processed_mb instead of the iterator?

context: https://github.com/NVIDIA-NeMo/RL/pull/1757/changes#r2757485690

**Review by terrykong — COMMENTED** [review]
thanks for all the refactoring work @ashors1 

1. I think we should merge this after @asolergi-nv looks into the padding issue in the packing path, since it would be good to check if this PR doesn't introduce a regression there

cc @ananthsub

File: nemo_rl/models/megatron/train.py (hunk 5 above)
Comment (yuki-97) [reply nemo_rl/models/megatron/train.py]: +1, and shall we move this function to a common place (maybe base_policy_worker) in a later PR?

cc @hemildesai

File: nemo_rl/models/policy/workers/megatron_policy_worker.py
Comment (yuki-97) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: not introduced by this PR, but is it useless since we will do `self.model.zero_grad_buffer()` at the beginning of each batch?

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 9]
```diff
@@ -343,63 +915,56 @@ def train(
...
-                    batch,
-                    self.cfg,
-                    mbs,
-                    straggler_timer=self.mcore_state.straggler_timer,
-                )
+                ) = get_microbatch_iterator(batch, self.cfg, mbs)
                 # Track total microbatches for MoE aux-loss averaging
                 total_num_microbatches += int(num_microbatches)
 
+                loss_fn_wrapped = LossPostProcessor(
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: nit: how about keep a same name with DTensor?
```suggestion
                loss_post_processor = LossPostProcessor(
```
and also below `post_processing_fn=loss_post_processor,`

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 10]
```diff
@@ -455,19 +1020,15 @@ def train(
...
-                    torch.distributed.broadcast_object_list(
-                        loss_metrics,
-                        src=get_pipeline_model_parallel_last_rank(),
-                        group=get_pipeline_model_parallel_group(),
-                    )
-                    gb_loss_metrics = loss_metrics[0]
+                    gb_loss_metrics = None
+
+                # Broadcast loss metrics from last stage to all stages
+                ## TODO: check with PP > 1
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py]: is it a debug thing forget to remove or a pending thing need to check?

File: nemo_rl/models/policy/workers/megatron_policy_worker.py
Comment (yuki-97) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:347]: shall we unify `empty_unused_memory_level` here and `clear_cache_every_n_steps` in DTensor?
https://github.com/NVIDIA-NeMo/RL/pull/1757#discussion_r2759333901

Comment (yuki-97) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:422]: do you think it's better to move the statistic part to a function as well?
https://github.com/NVIDIA-NeMo/RL/pull/1757#discussion_r2759435917

**Review by yuki-97 — COMMENTED** [review]
thank you so much for the refactor efforts! just one last step to complete. 🎉

File: nemo_rl/models/policy/workers/megatron_policy_worker.py
Comment (yuki-97) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:347]: as discussed offline, no changes need at megatron side.

File: nemo_rl/models/megatron/data.py [hunk 11]
```diff
@@ -294,15 +296,15 @@ def process_global_batch(
     *,
     batch_idx: int,
     batch_size: int,
-) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+) -> dict[str, Any]:
```
Comment (ananthsub) [inline nemo_rl/models/megatron/data.py:299]: return typehint wasn't in sync with the code

File: nemo_rl/models/megatron/data.py [hunk 12]
```diff
@@ -294,15 +296,15 @@ def process_global_batch(
...
     """Process a global batch and compute normalization factors.
 
     Args:
-        data: Full dataset
+        data: Full dataset to extract a batch from
+        loss_fn: Loss function (used to check loss type for token-level validation)
+        dp_group: Data parallel process group for all-reduce
```
Comment (ananthsub) [inline nemo_rl/models/megatron/data.py:305]: nit: docstring to match argument order

File: nemo_rl/models/megatron/data.py [hunk 13]
```diff
@@ -221,7 +222,8 @@ def process_microbatch(
     Optional[torch.Tensor],
 ]:
     """Process a microbatch for Megatron model forward pass."""
-    with straggler_timer(bdata=True):
+    ctx = straggler_timer(bdata=True) if straggler_timer is not None else nullcontext()
```
Comment (ananthsub) [inline nemo_rl/models/megatron/data.py:225]: handle straggler detection conditionally being passed

File: nemo_rl/models/megatron/train.py (hunk 8 above)
Comment (ananthsub) [reply nemo_rl/models/megatron/train.py:122]: it's harder to pass the processed_mb instead of the iterator in order to be compatible with the mcore pipeline parallel signature requirement. `forward_with_post_processing_fn` is passed as the forward step func in L244-264 which requires the `forward_step_func(data_iterator, model)` signature

File: nemo_rl/models/policy/workers/megatron_policy_worker.py
Comment (yuki-97) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:422]: can refer to the `aggregate_training_statistics ` in DTensor.
https://github.com/NVIDIA-NeMo/RL/blob/ed718b47d9469e9ddfc28876657f12956003247e/nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py#L497-L504

**Review by terrykong — APPROVED** [review]

**Comment by yuki-97** [comment]
cancelled CI for now and waiting https://github.com/NVIDIA-NeMo/RL/pull/1902 to be merged first.
